readers/mutation_reader.cc

```diff
@@ -112,38 +95,97 @@ bool mutation_fragment_stream_validator::operator()(mutation_fragment_v2::kind k
             valid = kind == mutation_fragment_v2::kind::partition_start;
             break;
     }
-    if (valid) {
-        _prev_kind = kind;
+    if (!valid) {
+        return false;
+    }
+
+    if (pos && _prev_kind != mutation_fragment_v2::kind::partition_end) {
+        auto cmp = position_in_partition::tri_compare(_schema);
+        auto res = cmp(_prev_pos, *pos);
+        if (_prev_kind == mutation_fragment_v2::kind::range_tombstone_change) {
+            valid &= res <= 0;
+        } else {
+            valid &= res < 0;
+        }
+        if (!valid) {
+            return false;
+        }
     }
-    return valid;
+
+    _prev_kind = kind;
+    if (pos) {
+        _prev_pos = *pos;
+    } else {
+        switch (kind) {
+            case mutation_fragment_v2::kind::partition_start:
+                _prev_pos = position_in_partition(position_in_partition::partition_start_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::static_row:
+                _prev_pos = position_in_partition(position_in_partition::static_row_tag_t{});
+                break;
+            case mutation_fragment_v2::kind::clustering_row:
+            case mutation_fragment_v2::kind::range_tombstone_change: // fall-through
```

Comment:
nit: use `[[fallthrough]];` on the case above (that falls) through, if at all.
https://en.cppreference.com/w/cpp/language/attributes/fallthrough
That said, I'm not sure this is really needed for a classic sequence of case labels like here.